<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>MK approaches · MKtest.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://jmurga.github.io/MKtest.jl/mk/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MKtest.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Package Overview</span><ul><li><a class="tocitem" href="../empirical/">Infering the rate and strength of adaptation</a></li><li><a class="tocitem" href="../rates/">Rates</a></li><li><a class="tocitem" href="../summstat/">Summary statistics</a></li><li><a class="tocitem" href="../abc/">ABC inference</a></li></ul></li><li><a class="tocitem" href="../cli/">Command-Line interface</a></li><li class="is-active"><a class="tocitem" href>MK approaches</a><ul class="internal"><li><a class="tocitem" href="#Standard-MKT"><span>Standard MKT</span></a></li><li><a class="tocitem" href="#Fay,-Waycoff-and-Wu-MK-extension"><span>Fay, Waycoff and Wu MK extension</span></a></li><li><a class="tocitem" href="#imputed-MKT-(in-preparation)"><span>imputed MKT (in preparation)</span></a></li><li><a class="tocitem" href="#Asymptotic-MKT"><span>Asymptotic MKT</span></a></li><li><a class="tocitem" href="#Grapes"><span>Grapes</span></a></li></ul></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>MK approaches</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>MK approaches</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jmurga/MKtest.jl/blob/master/docs/src/mk.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="MK-approaches"><a class="docs-heading-anchor" href="#MK-approaches">MK approaches</a><a id="MK-approaches-1"></a><a class="docs-heading-anchor-permalink" href="#MK-approaches" title="Permalink"></a></h1><p>We included other MK approaches in our module. All the functions use the formated SFS and divergence data described at <a href="../summstat/">Summary statistics section</a>.</p><h2 id="Standard-MKT"><a class="docs-heading-anchor" href="#Standard-MKT">Standard MKT</a><a id="Standard-MKT-1"></a><a class="docs-heading-anchor-permalink" href="#Standard-MKT" title="Permalink"></a></h2><p>The standard McDonald &amp; Kreitman test ([McDonald and Kreitman, 1991]) was developed to be applied to protein coding sequences, combining both divergence (<span>$D$</span>) and polymorphism (<span>$P$</span>) sites, and categorizing mutations as synonymous (<span>$P_S$</span>, <span>$D_S$</span>) and non-synonymous (<span>$P_N$</span>, <span>$D_N$</span>). </p><p>If all mutations are either strongly deleterious or neutral, then <span>$Di/D0$</span> is expected to roughly equal <span>$Pi/P0$</span>. In contrast, if positive selection is operating in the region, adaptive mutations rapidly reach fixation and contribute more to divergence than polymorphism compared to neutral mutations, and then <span>$D_N/D_S &gt; P_N/P_S$</span>. Assuming that adaptive mutations contribute little to polymorphism but substantially to divergence, the proportion of non-synonymous substitutions is inferred following <a href="https://doi.org/10.1038/4151022a">Smith and Eyre-Walker (2002)</a>.</p><p class="math-container">\[\alpha = 1 - (\frac{P_N}{P_S}\cdot\frac{D_S}{D_N})\]</p><p>Please check <a href="../reference/#MKtest.standardMK"><code>MKtest.standardMK</code></a> to obtain more info.</p><pre><code class="language-julia hljs">adap = MKtest.parameters(n=661,cutoff=[0.0,1.0])
alpha, sfs, divergence, m= MKtest.parse_sfs(adap, data = &quot;mktest/tgp.txt&quot;)

mk = MKtest.standardMK(sfs[1],divergence[1])</code></pre><h2 id="Fay,-Waycoff-and-Wu-MK-extension"><a class="docs-heading-anchor" href="#Fay,-Waycoff-and-Wu-MK-extension">Fay, Waycoff and Wu MK extension</a><a id="Fay,-Waycoff-and-Wu-MK-extension-1"></a><a class="docs-heading-anchor-permalink" href="#Fay,-Waycoff-and-Wu-MK-extension" title="Permalink"></a></h2><p><a href="[fwwMK](https:/doi.org/10.1038/4151024a)">Fay et al. (2002)</a> proposed an approach that removes all polymorphisms segregating at a frequency (<span>$j$</span>) below a given threshold (usually <span>$j &gt; 5\%–15\%$</span>). Although there is no consensus about what this threshold should be used, <a href="https://doi.org/10.1093/molbev/msn005">J. Charlesworth &amp; Eyre-Walker (2008)</a> demonstrated that  estimates are robust using a frequency threshold of 15%, below which most slightly deleterious polymorphisms are found and removed. The estimates are reasonably accurate only when the rate of adaptive evolution is high and the Distribution of Fitness Effects (DFE) of deleterious mutations is leptokurtic.</p><p class="math-container">\[\alpha = 1 - (\frac{P_{N (j&gt;15\%)}}{P_{S (j&gt;15\%)}\cdot\frac{D_S}{D_N})\]</p><p>Please check <a href="../reference/#MKtest.fwwMK"><code>MKtest.fwwMK</code></a> to obtain more info.</p><pre><code class="language-julia hljs">fww = MKtest.fwwMK(adap,sfs[1],divergence[1],cutoff=0.15)</code></pre><h2 id="imputed-MKT-(in-preparation)"><a class="docs-heading-anchor" href="#imputed-MKT-(in-preparation)">imputed MKT (in preparation)</a><a id="imputed-MKT-(in-preparation)-1"></a><a class="docs-heading-anchor-permalink" href="#imputed-MKT-(in-preparation)" title="Permalink"></a></h2><p>The imputed MKT (impMKT) is a modification of the Fay, Waycoff, and Wu MK extension (fwwMK) (<a href="[fwwMK](https:/doi.org/10.1038/4151024a)">Fay et al. (2002)</a> to improve gene-by-gene analyses. The method propose the imputation of slightly deleterious mutations at the SFS rather than removing all variants below a frequency threshold. The imputedMK aims to maximize the information to test the excess of divergence ratio relative to polymorphism at the gene level. <span>$\alpha$</span> is estimated as</p><p class="math-container">\[\alpha_{imputed} = 1 - \left( \frac{P_{N} - P_{wd}}{P_{S}} \cdot \frac{D_{N}}{D_{S}} \right)\]</p><p>where <span>$P_{wd}$</span> is</p><p class="math-container">\[P_{wd} \approx P_{wd (j &lt; 15\%)} = P_{N (j&lt;15\%)} - \frac{P_{N (j&gt;15\%) } \cdot P_{S (j&lt;15\%)}}{P_{S (j&gt;15\%)}}\]</p><p>Please check <a href="@ref"><code>MKtest.imputedMK</code></a> to obtain more info.</p><pre><code class="language-julia hljs">imputed = MKtest.imputedMK(adap,sfs[1],divergence[1],cutoff=0.15)</code></pre><h2 id="Asymptotic-MKT"><a class="docs-heading-anchor" href="#Asymptotic-MKT">Asymptotic MKT</a><a id="Asymptotic-MKT-1"></a><a class="docs-heading-anchor-permalink" href="#Asymptotic-MKT" title="Permalink"></a></h2><p>Proposed by Messer and Petrov (2013). This extension is robust to the presence of selective sweeps (genetic hitchhiking) and the segregation of slightly deleterious polymorphisms substitutions (BGS). In this approach, the authors defined <span>$\alpha$</span> as a function that depends on the SFS of alleles. Therefore, <span>$\alpha$</span> is estimated in different frequency intervals (<span>$x$</span>), and these values are then adjusted to an exponential function. An exponential fit is suitable as the non-synonymous allele frequency is expected to decay exponentially over the respective levels of synonymous polymorphisms (Messer &amp; Petrov, 2013). <span>$\alpha$</span> is estimated as</p><p class="math-container">\[\alpha_{fit(x)} = a+b \cdot e^{-cx}\]</p><p>Please check <a href="../reference/#MKtest.aMK"><code>MKtest.aMK</code></a> to obtain more info.</p><pre><code class="language-julia hljs">amk, ci, model = MKtest.aMK(adap,alpha[1])</code></pre><h2 id="Grapes"><a class="docs-heading-anchor" href="#Grapes">Grapes</a><a id="Grapes-1"></a><a class="docs-heading-anchor-permalink" href="#Grapes" title="Permalink"></a></h2><p>Grapes is a ML method that can estimate the expected proportion of adaptive fixations given the inferred DFE from the MK data. Grapes assumes can model the DFE in the form of two different versions of the Fisher&#39;s geometric model, and a model assuming a Beta-shaped distribution of weak effect mutations, a Gamma distribution as well as a exponential distibution. Please check (Grapes repository)[https://github.com/BioPP/grapes] a cite (Galtier 2016)[https://doi.org/10.1371/journal.pgen.1005774] if you use <code>MKtest.grapes</code> function.</p><pre><code class="language-julia hljs">grapes_df = grapes(sfs,divergence,m,&quot;GammaExpo&quot;,&quot;mktest/&quot;,20)</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../cli/">« Command-Line interface</a><a class="docs-footer-nextpage" href="../bootstrap/">Bootstrap »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.13 on <span class="colophon-date" title="Wednesday 20 September 2023 14:04">Wednesday 20 September 2023</span>. Using Julia version 1.9.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
