<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Infering the rate and strength of adaptation · MKtest.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://jmurga.github.io/MKtest.jl/empirical/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MKtest.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Package Overview</span><ul><li class="is-active"><a class="tocitem" href>Infering the rate and strength of adaptation</a><ul class="internal"><li><a class="tocitem" href="#Extending-analytical-estimations"><span>Extending analytical estimations</span></a></li><li class="toplevel"><a class="tocitem" href="#Computational-pipeline"><span>Computational pipeline</span></a></li></ul></li><li><a class="tocitem" href="../rates/">Rates</a></li><li><a class="tocitem" href="../summstat/">Summary statistics</a></li><li><a class="tocitem" href="../abc/">ABC inference</a></li></ul></li><li><a class="tocitem" href="../cli/">Command-Line interface</a></li><li><a class="tocitem" href="../mk/">MK approaches</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Package Overview</a></li><li class="is-active"><a href>Infering the rate and strength of adaptation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Infering the rate and strength of adaptation</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jmurga/MKtest.jl/blob/master/docs/src/empirical.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><p>We developed an extension of the analytical approximation presented <a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. (2019)</a>. In such a paper, the authors used the analytical approximation to <span>$\alpha_x$</span> to explore the effect of BGS on weakly beneficial alleles, but the empirical estimation of <span>$\alpha$</span> procedure employed was based on computationally intensive forward-in-time simulations. Unfortunately, forward-in-time simulations often require high-performance computing resources and can be prohibitively slow for some models. Here we extend <a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. (2019)</a>analytical approximations to develop a simple and computationally efficient ABC-based inference procedure modelling the DFE of deleterious and beneficial alleles and incomplete recombination between selected genomic elements. </p><p>We followed a generic ABC scheme (<a href="https://doi.org/10.1093/genetics/162.4.2025">Beaumont et al. (2002)</a>) to perform the empirical estimation of <span>$\alpha$</span>. ABC scheme proceeds by first sampling parameter values from prior distributions, simulating the model using these parameter values, calculating informative summary statistics, and comparing the simulated summary statistics to the observed data. Following the standard ABC scheme, we 1) considered <span>$N$</span> random combinations from prior distributions; 2) solved <span>$N$</span> independent models to generate informative summary statistics; 3) subset the parameter values producing summary statistics that best match the observed data from an approximate posterior distribution. Additionally, a linear model can be imposed to correct the non-0 distance between the simulated and observed summary statistics.</p><h2 id="Extending-analytical-estimations"><a class="docs-heading-anchor" href="#Extending-analytical-estimations">Extending analytical estimations</a><a id="Extending-analytical-estimations-1"></a><a class="docs-heading-anchor-permalink" href="#Extending-analytical-estimations" title="Permalink"></a></h2><p><a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. (2019)</a> used the analytical theory to demonstrate the effect of weakly beneficial alleles at <span>$\alpha_{(x)}$</span>. To do that, they solved <span>$\alpha_{(x)}$</span> through the fixation and polymorphism rates since the locus length (<span>$L$</span>) and the time branch (<span>$T$</span>) estimations at first order <span>$\alpha_{(x)}$</span> estimation can be canceled</p><p class="math-container">\[\mathbb{E}[\alpha_{(x)}] \approx 1 - \frac{LT(\mathbb{E}[d_S])}{LT(\mathbb{E}[d_N])} \frac{LT(\mathbb{E}[p_{S(x)}])}{LT(\mathbb{E}[p_{N(x)}])} \approx 1 - \frac{\mathbb{E}[d_S]}{\mathbb{E}[d_N]} \frac{\mathbb{E}[p_{S(x)}]}{\mathbb{E}[p_{N(x)}]}\]</p><p>where <span>$\frac{LT}{LT}$</span> with a constant mutation rate tend to <span>$1$</span>. </p><p>Here, we extended their analytical calculations to generate the summary statistics required at ABC approaches. Thereby we avoided expensive forward simulations. </p><p>Considering fixed values of <span>$T$</span>, <span>$L$</span>, <span>$\mu$</span>, and fixation rates it is possible to solve analytical <span>$\alpha_{(x)}$</span> performing the multiplication. However, this requires not only branch length estimations but also explicitly locus length selection which highly increases the order of estimations to solve. To avoid performing branch length estimations in our computation, we assumed that the empirically observed number of fixations should be proportional to the length of the evolutionary branch of interest, <span>$T$</span>, the locus length <span>$L$</span> and mutation ration <span>$\mu$</span>. We take the observed total number of fixations (including both nonsynonymous and synonymous sites) as a proxy for the expected number, and then sample weakly deleterious, neutral, and beneficial substitutions proportional to their relative rates for a fixed set of model parameters. The expected number of substitutions for positively selected substitutions is then</p><p class="math-container">\[\mathbb{E}[D] = X \in Poisson\left(\lambda = D_{observed} \times (\mathbb{E}[d_+]+\mathbb{E}[d_-]+\mathbb{E}[d_0])\right)\]</p><p>To estimate <span>$\alpha_{(x)}$</span> we draw the expected values considering each fixation category. We sampled the values in two categories according to the relative rates. Following the same procedure:</p><p class="math-container">\[\mathbb{E}[D_S] = X \in Poisson\left(\lambda = D_{observed} \times \left[\frac{\mathbb{E}[d_0]}{\mathbb{E}[d_+] + \mathbb{E}[d_-] + \mathbb{E}[d_0]}\right]\right)\]</p><p class="math-container">\[\mathbb{E}[D_N] = X \in Poisson\left(\lambda = D_{observed} \times \left[\frac{\mathbb{E}[d_+] + \mathbb{E}[d_-]}{\mathbb{E}[d_+] + \mathbb{E}[d_-] + \mathbb{E}[d_0]}\right]\right)\]</p><p>It should be noted that both sampling variance and process variance affect the number of variable alleles at any particular allele frequency in a sequencing sample. </p><p>To incorporate that variance we made one sample per frequency-bin given the SFS, we sampled a Poisson distributed number of polymorphic alleles at frequency <span>$x$</span> relative to their rate given the expected frequency spectra. The expected frequency spectra were downsampled using a binomial (with probability of success given by the frequency <span>$\begin{pmatrix} x \\ 2n \end{pmatrix}$</span> in a sample of <span>$2n$</span> chromosomes) to account for the sampling variance. Considering the random process at each frequency we made credible intervals for future ABC estimations. Otherwise, we would falsely find higher confidence in our parameter estimates. We draw the expected polymorphic values similarly to fixations, considering the SFS and the expected rates.</p><p class="math-container">\[\mathbb{E}[P] = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)\ observed} \times (\mathbb{E}[p_{+(x)}]+\mathbb{E}[p_{-(x)}]+\mathbb{E}[p_{0(x)}])\right)\]</p><p>It should be noted that we modified the expected SFS and the observed SFS considering each frequency <span>$x$</span> as the cumulative sum above <span>$x$</span>. This quantities have the same asymptote but are less affected by changing sample size. Therefore, we scaled better the analysis to sample size since most common alleles at frequencies <span>$x$</span> have very few polymorphic sites in large samples. This way, we finally transformed <span>$\alpha_{(x)}$</span> to be depending on the previous frequency bin value, which reflects each frequency category over the expected asymptotic shape. Because of that, <span>$\alpha_{(x)}$</span> analysis is more robust even in low polymorphic populations.</p><p>We followed the same procedure described above to subset the expected number of fixation taking into account weakly, strong and deleterious fixation rates categories. </p><p class="math-container">\[\mathbb{E}[D_W] = X \in Poisson\left(\lambda = D_{observed} \times \left[\frac{\mathbb{E}[d_W]}{\mathbb{E}[d_+] + \mathbb{E}[d_-] + \mathbb{E}[d_0]}\right]\right)\]</p><p class="math-container">\[\mathbb{E}[D_S] = X \in Poisson\left(\lambda = D_{observed} \times \left[\frac{\mathbb{E}[d_S]}{\mathbb{E}[d_+] + \mathbb{E}[d_-] + \mathbb{E}[d_0]}\right]\right)\]</p><p>In addition, for each model, we retrieve deleterious DFE parameters to perform ABC inference.</p><h1 id="Computational-pipeline"><a class="docs-heading-anchor" href="#Computational-pipeline">Computational pipeline</a><a id="Computational-pipeline-1"></a><a class="docs-heading-anchor-permalink" href="#Computational-pipeline" title="Permalink"></a></h1><p>The following sections describe a pipeline to estimate the empirical adaptation rate (<span>$\alpha$</span>) using whole-genome protein, Viral Interacting Proteins and Non-Viral Interacting Protein data from the TGP. This analysis is similar to the one performed at <a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. (2019)</a>. We divided the pipeline into the following step:</p><ul><li>Expected fixations rates and frequency spectra estimations</li><li>Summary statistic estimations</li><li>ABC inference</li></ul><p>Please note that we use a relative path called <em>mktest/</em> to execute the whole pipeline. This path will be relative to the where Julia or the Command-Line Interface is executed.</p><p>The software is prepared to parallelize each pipeline step using Julia using multi-threading computing. You can add the desired number of threads executing <code>-t</code> option when running julia (e.g using 8 cores: <code>julia -t8</code>). Please make some tests in your machine before executing expensive models. It is almost mandatory to parallelize at least the rate estimations.</p><p>The following examples were tested using a laptop with the following hardware:</p><ul><li>Intel i7-7700HQ (8) @ 3.800GHz </li><li>16GB RAM DDR4 2400MHz</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../rates/">Rates »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.13 on <span class="colophon-date" title="Sunday 7 January 2024 10:44">Sunday 7 January 2024</span>. Using Julia version 1.10.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
