<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reference · MKtest.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://jmurga.github.io/MKtest.jl/reference/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MKtest.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Package Overview</span><ul><li><a class="tocitem" href="../empirical/">Infering the rate and strength of adaptation</a></li><li><a class="tocitem" href="../rates/">Rates</a></li><li><a class="tocitem" href="../summstat/">Summary statistics</a></li><li><a class="tocitem" href="../abc/">ABC inference</a></li></ul></li><li><a class="tocitem" href="../cli/">Command-Line interface</a></li><li><a class="tocitem" href="../mk/">MK approaches</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li class="is-active"><a class="tocitem" href>Reference</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#MKtest-estimation"><span>MKtest estimation</span></a></li><li><a class="tocitem" href="#Fixations"><span>Fixations</span></a></li><li><a class="tocitem" href="#Polymorphism"><span>Polymorphism</span></a></li><li><a class="tocitem" href="#Rates"><span>Rates</span></a></li><li><a class="tocitem" href="#Summary-statistics"><span>Summary statistics</span></a></li><li><a class="tocitem" href="#Inference-tools"><span>Inference tools</span></a></li><li><a class="tocitem" href="#MKtest"><span>MKtest</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Reference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jmurga/MKtest.jl/blob/master/docs/src/reference.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Model-parameters"><a class="docs-heading-anchor" href="#Model-parameters">Model parameters</a><a id="Model-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Model-parameters" title="Permalink"></a></h1><p><em>adap</em> is the only variable exported from <em>MKtest</em> module. It is a Mutable structure contaning the variables required to solve the analytical approach. Any value can be easly changed. Remember <em>adap</em> should be change before the execution, in other case, <span>$\alpha_{(x)}$</span> will be solve with the default values. To change all the values at once, you can use <a href="#MKtest.parameters"><code>MKtest.parameters</code></a> in order to set specific models in a new variable.</p><article class="docstring"><header><a class="docstring-binding" id="MKtest.parameters" href="#MKtest.parameters"><code>MKtest.parameters</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Mutable structure containing the variables required to solve the analytical approach. All the functions are solve using the internal values of the structure. You should declare a mutable structure to the perform the analytical estimations.</p><p><strong>Parameters</strong></p><ul><li><code>gam_flanking::Int64</code>: Selection coefficient for deleterious alleles</li><li><code>gL::Int64</code>: Selection coefficient for weakly benefical alleles</li><li><code>gH::Int64</code>: Selection coefficient for strongly benefical alleles</li><li><code>al_low::Float64</code>: Proportion of α due to weak selection</li><li><code>al_tot::Float64</code>: α</li><li><code>θ_flanking::Float64</code>: Mutation rate defining BGS strength</li><li><code>θ_coding::Float64</code>: Mutation rate on coding region</li><li><code>al::Float64</code>: DFE shape parameter </li><li><code>be::Float64</code>: DFE scale parameter</li><li><code>B::Float64</code>: BGS strength</li><li><code>B_bins::Array{Float64,1}</code>: BGS values to simulate</li><li><code>ppos_l::Float64</code>: Fixation probabily of weakly beneficial alleles</li><li><code>ppos_h::Float64</code>: Fixation probabily of strongly beneficial alleles</li><li><code>N::Int64</code>: Population size</li><li><code>n::Int64</code>: Sample size</li><li><code>Lf::Int64</code>: Flanking region length</li><li><code>ρ::Float64</code>: Recombination rate</li><li><code>TE::Float64</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/parameters.jl#L5-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.analytical_alpha" href="#MKtest.analytical_alpha"><code>MKtest.analytical_alpha</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">analytical_alpha(param)</code></pre><p>Analytical α(x) estimation. Solve α(x) generally. We used the expected fixation rates and SFSto approach the asympotic value accouting for background selection, weakly and strong positive selection. α(x) can be estimated taking into account the role of positive selected alleles or not. In this way we explore the role of linkage to deleterious alleles in the coding region.</p><p class="math-container">\[\mathbb{E}[\alpha_{x}] =  1 - \left(\frac{\mathbb{E}[D_{s}]}{\mathbb{E}[D_{N}]}\frac{\mathbb{E}[P_{N}]}{\mathbb{E}[P_{S}]}\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{Vector{Float64},Vector{Float64}}</code>: α(x) accounting for weak adaptation, α(x) non-accouting for weak adaptation.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/parameters.jl#L264-L277">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.Br" href="#MKtest.Br"><code>MKtest.Br</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Br(Lmax,theta)</code></pre><p>Expected reduction in nucleotide diversity. Explored at <a href="https://doi.org/10.1017/S0016672300032365">Charlesworth B., 1994</a>:</p><p class="math-container">\[\frac{\pi}{\pi_{0}} = e^{\frac{4\mu L}{2rL+t}}\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li><li><code>theta::Float64</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected reduction in diversity given a non-coding length, mutation rate and defined recombination.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/parameters.jl#L74-L88">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.set_θ</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="MKtest.binom_op" href="#MKtest.binom_op"><code>MKtest.binom_op</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">binom_op(param)</code></pre><p>Binomial convolution to sample the allele frequencies probabilites depending on background selection values and sample size.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li><li><code>B_bins::Vector{Float64}</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Float64,2}</code>: convoluted SFS for each B value defined in the model (B_bins). The estimations are saved at <em>convolutedBn.bn</em>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/parameters.jl#L167-L178">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.Φ" href="#MKtest.Φ"><code>MKtest.Φ</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Φ(param,s)</code></pre><p>Reduction in fixation probabilty due to background selection and linkage. The formulas used have been subjected to several theoretical works (<a href="https://doi.org/10.1017/S0016672300032365">Charlesworth B., 1994</a>, <a href="https://www.genetics.org/content/141/4/1605">Hudson et al., 1995</a>, <a href="https://doi.org/10.1017/S0016672300033619">Nordborg et al. 1995</a>, <a href="https://www.genetics.org/content/140/2/821">Barton NH., 1995</a>).</p><p>The fixation probabilty of selected alleles are reduce by a factor <span>$\phi$</span>:</p><p class="math-container">\[\phi(t,s) = e^{[\frac{-2\mu}{t(1+\frac{rL}{t}+\frac{2s}{t})}]}\]</p><p>Multiplying across all deleterious linkes sites, we find:</p><p class="math-container">\[\Phi = \prod_{1}^{L} = \phi(t,s) = e^{\frac{-2t\mu(\psi[1,\frac{r+2s+L}{r}] - \psi[1,\frac{r(L+1)+2s+t}{r}])}{r^2}}\]</p><p class="math-container">\[\phi(t,s) = e^{\frac{-2t\mu(\psi[1,\frac{r+2s+L}{r}] - \psi[1,\frac{r(L+1)+2s+t}{r}])}{r^2}}\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model</li><li><code>gamma::Int64</code>: selection coefficient.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixations under background selection.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/parameters.jl#L219-L249">source</a></section></article><h1 id="MKtest-estimation"><a class="docs-heading-anchor" href="#MKtest-estimation">MKtest estimation</a><a id="MKtest-estimation-1"></a><a class="docs-heading-anchor-permalink" href="#MKtest-estimation" title="Permalink"></a></h1><h2 id="Fixations"><a class="docs-heading-anchor" href="#Fixations">Fixations</a><a id="Fixations-1"></a><a class="docs-heading-anchor-permalink" href="#Fixations" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.fix_neut" href="#MKtest.fix_neut"><code>MKtest.fix_neut</code></a> — <span class="docstring-category">Function</span></header><section><div><ul><li>fix_neut(parameters)</li></ul><p>Expected neutral fixations rate reduce by B value.</p><p class="math-container">\[\mathbb{E}[D_{s}] = (1 - p_{-} - p_{+}) B \frac{1}{2N}\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of neutral fixations.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/fixations.jl#L8-L24">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.fix_neg" href="#MKtest.fix_neg"><code>MKtest.fix_neg</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fix_neg(parameters,ppos)</code></pre><p>Expected fixation rate from negative DFE.</p><p class="math-container">\[\mathbb{E}[D_{n-}] =  p_{-}\left(2^-\alpha\beta^\alpha\left(-\zeta[\alpha,\frac{2+\beta}{2}] + \zeta[\alpha,1/2(2-\frac{1}{N+\beta})]\right)\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model.</li><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of fixations from negative DFE.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/fixations.jl#L32-L49">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.p_fix" href="#MKtest.p_fix"><code>MKtest.p_fix</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">p_fix(parameters,gam)</code></pre><p>Expected positive fixation rate.</p><p class="math-container">\[\mathbb{E}[D_{n+}] =  p_{+} \cdot B \cdot (1 - e^{(-2s)})\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model.</li><li><code>gam::Int64</code>: positive selection coefficient.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixation.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/fixations.jl#L68-L84">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.fix_pos_sim" href="#MKtest.fix_pos_sim"><code>MKtest.fix_pos_sim</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fix_pos_sim(parameters,gamma,ppos)</code></pre><p>Expected positive fixations rate reduced due to the impact of background selection and linkage. The probabilty of fixation of positively selected alleles is reduced by a factor Φ across all deleterious linked sites <a href="@ref"><code>Analytical.Φ</code></a>.</p><p class="math-container">\[\mathbb{E}[D_{n+}] =  \Phi \cdot \mathbb{E}[D_{n+}]\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model.</li><li><code>gam</code>::Int64: positive selection coefficient.</li><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixations under background selection</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/fixations.jl#L106-L124">source</a></section></article><h2 id="Polymorphism"><a class="docs-heading-anchor" href="#Polymorphism">Polymorphism</a><a id="Polymorphism-1"></a><a class="docs-heading-anchor-permalink" href="#Polymorphism" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.sfs_neut" href="#MKtest.sfs_neut"><code>MKtest.sfs_neut</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sfs_neut(param,binom)</code></pre><p>Expected rate of neutral allele frequency reduce by background selection. The spectrum depends on the number of individual []</p><p class="math-container">\[\mathbb{E}[Ps_{(x)}] = \sum{x^{*}=x}{x^{*}=1}f_{B}(x)\]</p><p><strong>Input</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model.</li><li><code>SparseMatrixCSC{Float64, Int64}</code>: convoluted SFS given the B value.</li></ul><p><strong>Return:</strong></p><ul><li><code>Vector{Float64}</code>: expected rate of neutral alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/polymorphism.jl#L10-L24">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.sfs_pos" href="#MKtest.sfs_pos"><code>MKtest.sfs_pos</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sfs_pos(param,s,p,binom)</code></pre><p>Expected rate of positive selected allele frequency reduce by background selection. The spectrum depends on the number of individuals.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model.</li><li><code>s::Int64</code>: selection coefficient.</li><li><code>p::Float64</code>: positive selected alleles probabilty.</li><li><code>binom::SparseMatrixCSC{Float64, Int64}</code>: convoluted SFS given the B value.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected positive selected alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/polymorphism.jl#L40-L53">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.sfs_neg" href="#MKtest.sfs_neg"><code>MKtest.sfs_neg</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sfs_neg(param,p,binom)</code></pre><p>Expected rate of positive selected allele frequency reduce by background selection. Spectrum drawn on a gamma DFE. It depends on the number of individuals.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model</li><li><code>p::Float64</code>: positive selected alleles probabilty.</li><li><code>binom::SparseMatrixCSC{Float64, Int64}</code>: convoluted SFS given the B value.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected negative selected alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/polymorphism.jl#L143-L155">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.cumulative_sfs</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="MKtest.reduce_sfs" href="#MKtest.reduce_sfs"><code>MKtest.reduce_sfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">reduce_sfs(sfs_tmp,bins)</code></pre><p>Function to reduce the SFS into N bins.</p><p><strong>Arguments</strong></p><ul><li><code>sfs_tmp::Vector{Float64}</code>: SFS vector.</li><li><code>bins::Int64</code>: bin size to collapse the SFS.</li></ul><p><strong>Output</strong></p><ul><li><code>Vector{Float64}</code>: Binned SFS vector.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/polymorphism.jl#L242-L252">source</a></section></article><h2 id="Rates"><a class="docs-heading-anchor" href="#Rates">Rates</a><a id="Rates-1"></a><a class="docs-heading-anchor-permalink" href="#Rates" title="Permalink"></a></h2><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.rates</code>. Check Documenter&#39;s build log for details.</p></div></div><h2 id="Summary-statistics"><a class="docs-heading-anchor" href="#Summary-statistics">Summary statistics</a><a id="Summary-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Summary-statistics" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.poisson_fixation" href="#MKtest.poisson_fixation"><code>MKtest.poisson_fixation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">poisson_fixation(observed_values,λds, λdn)</code></pre><p>Divergence sampling from Poisson distribution. The expected neutral and selected fixations are subset through their relative expected rates (<a href="#MKtest.fix_neut"><code>fix_neut</code></a>, <a href="@ref"><code>fix_neg_b</code></a>, <a href="#MKtest.fix_pos_sim"><code>fix_pos_sim</code></a>). Empirical values are used are used to simulate the locus <em>L</em> along a branch of time <em>T</em> from which the expected <em>Ds</em> and <em>Dn</em> raw count estimated given the mutation rate (<span>$\mu$</span>). Random number generation is used to subset samples arbitrarily given the success rate <span>$\lambda$</span> in the distribution.</p><p class="math-container">\[\mathbb{E}[D_N] = X \in Poisson\left(\lambda = D \times \left[\frac{\mathbb{E}[D_+] + \mathbb{E}[D_-]}{\mathbb{E}[D_+] + \mathbb{E}[D_-] + \mathbb{E}[D_0]}\right]\right)\]</p><p class="math-container">\[\mathbb{E}[D_S] = X \in Poisson\left(\lambda = D \times \left[\frac{\mathbb{E}[D_0]}{\mathbb{E}[D_+] + \mathbb{E}[D_-] + \mathbb{E}[D_0]}\right]\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>observed_values::Array</code>: Array containing the total observed divergence.</li><li><code>λds::Float64</code>: expected neutral fixations rate.</li><li><code>λdn::Float64</code>: expected selected fixations rate.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Int64,1}</code> containing the expected count of neutral and selected fixations.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/summary_statistics.jl#L4-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.poisson_polymorphism" href="#MKtest.poisson_polymorphism"><code>MKtest.poisson_polymorphism</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">poisson_polymorphism(observed_values,λps,λpn)</code></pre><p>Polymorphism sampling from Poisson distributions. The total expected neutral and selected polimorphism are subset through the relative expected rates at the frequency spectrum (<a href="#MKtest.fix_neut"><code>fix_neut</code></a>, <a href="#MKtest.sfs_neut"><code>sfs_neut</code></a>,). Empirical sfs are used to simulate the locus <em>L</em> along a branch of time <em>T</em> from which the expected <em>Ps</em> and <em>Pn</em> raw count are estimated given the mutation rate (<span>$\mu$</span>). Random number generation is used to subset samples arbitrarily from the whole sfs given each frequency success rate <span>$\lambda$</span> in the distribution.</p><p>The success rate managing the Poisson distribution by the observed count each frequency.  We considered both sampling variance and process variance is affecting the number of variable alleles we sample from SFS. This variance arises from the random mutation-fixation process along the branch. To incorporate this variance we do one sample per frequency-bin and use the total sampled variation and the SFS for the summary statistics.</p><p class="math-container">\[\mathbb{E}[P_N]   = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)} \times \left[\frac{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}]}{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}] + \mathbb{E}[P_{0(x)}]}\right]\right)\]</p><p class="math-container">\[\mathbb{E}[P_S] = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)} \times \left[\frac{\mathbb{E}[P_{0(x)}]}{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}] + \mathbb{E}[P_{0(x)}]}\right]\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>observed_values::Array{Int64,1}</code>: Array containing the total observed divergence.</li><li><code>λps::Array{Float64,1}</code>: expected neutral site frequency spectrum rate.</li><li><code>λpn::Array{Float64,1}</code>: expected selected site frequency spectrum rate.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Int64,2}</code> containing the expected total count of neutral and selected polymorphism.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/summary_statistics.jl#L69-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.sampling_summaries" href="#MKtest.sampling_summaries"><code>MKtest.sampling_summaries</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sampled_alpha(observed_values,λds, λdn)</code></pre><p>Ouput the expected values from the Poisson sampling process. Please check <a href="#MKtest.poisson_fixation"><code>poisson_fixation</code></a> and <a href="#MKtest.poisson_polymorphism"><code>poisson_polymorphism</code></a> to understand the samplingn process. α(x) is estimated through the expected values of Dn, Ds, Pn and Ps.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: Array containing the total observed divergence.</li><li><code>d::Array</code>: observed divergence.</li><li><code>afs::Array</code>: observed polymorphism.</li><li><code>λdiv::Array{Float64,2}</code>: expected fixations rate.</li><li><code>λdiv::Array{Float64,2}</code>: expected site frequency spectrum rates.</li></ul><p><strong>Returns</strong></p><p>α<em>summaries,exp</em>dn,exp<em>ds,exp</em>pn,exp_ps,ssAlpha</p><ul><li><p><code>Array{Int64,2}</code> containing α(x) values.</p></li><li><p><code>Array{Int64,1}</code> expected non-synonymous divergence.</p></li><li><p><code>Array{Int64,1}</code> expected synonymous divergence.</p></li><li><p><code>Array{Int64,1}</code> expected non-synonymous polymorphism.</p></li><li><p><code>Array{Int64,1}</code> expected synonymous polymorphism.</p></li><li><p><code>Array{Int64,1}</code> expected synonymous polymorphism.</p></li><li><p><code>Array{Int64,1}</code> expected synonymous polymorphism.</p></li><li><p><code>Array{Int64,2}</code> containing α(x) binned values.</p><p>sampling<em>summaries(gammaL,gammaH,ppos</em>l,ppos_h,observedData,nopos)</p></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/summary_statistics.jl#L135-L159">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.summary_statistics" href="#MKtest.summary_statistics"><code>MKtest.summary_statistics</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">summary_statistics(param,h5_file,sfs,divergence,output_folder,summstat_size)</code></pre><p>Estimate summary statistics using observed data and analytical rates. <em>output_folder</em> will be used to output summary statistics</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code> : Mutable structure containing the models</li><li><code>h5_file::String</code> : HDF5 containing solved models, fixation and polymorphic rates</li><li><code>sfs::Vector</code>: SFS data parsed from parse_sfs().</li><li><code>divergence::Vector</code>: divergence data from parse_sfs().</li><li><code>output_folder::String</code> : path to save summary statistics and observed data.</li><li><code>summstat_size::Int64</code> : number of summary statistics to sample.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/summary_statistics.jl#L203-L215">source</a></section></article><h2 id="Inference-tools"><a class="docs-heading-anchor" href="#Inference-tools">Inference tools</a><a id="Inference-tools-1"></a><a class="docs-heading-anchor-permalink" href="#Inference-tools" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.parse_sfs" href="#MKtest.parse_sfs"><code>MKtest.parse_sfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">parse_sfs(param;data,gene_list,sfs_columns,div_columns,l_columns,bins,isolines)</code></pre><p>Function to parse polymorphism and divergence by subset of genes. The input data is based on supplementary material described at <a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. 2019</a>. Please be sure the file is tabulated.</p><table><tr><th style="text-align: right">GeneId</th><th style="text-align: right">Pn</th><th style="text-align: right">DAF seppareted by commas</th><th style="text-align: right">Ps</th><th style="text-align: right">DAF separated by commas</th><th style="text-align: right">Dn</th><th style="text-align: right">Ds</th></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr></table><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model.</li><li><code>data::S</code>: path to polymorphic and divergence raw data.</li><li><code>gene_list:S</code>: path to gene list. </li><li><code>sfs_columns::Vector{Int64}</code>: column numbers corresponding to Pn and Ps frequencies. Default values [3,5]</li><li><code>div_columns::Vector{Int64}</code>: column numbers corresponding to Dn and Ds frequencies. Default values [6,7]</li><li><code>l_columns::Vector{Int64}</code>: column numbers corresponding to total number of non-synonymous and synonymous sites. Not require at data. Default values [7,8]</li><li><code>bins::Union{Nothing,Int64}</code>: size to bin the SFS independtly of the sample size.</li></ul><p><strong>Returns</strong></p><ul><li><code>Vector{Vector{Float64}}</code>: α_x values. Estimated using the cumulative SFS</li><li><code>Vector{Matrix{Float64}}</code>: Site Frequency Spectrum (non-cumulative)</li><li><code>Vector{Matrix{Int64}}</code>: Synonymous and non-synonymous divergence counts</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/parse.jl#L5-L28">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.ABCreg" href="#MKtest.ABCreg"><code>MKtest.ABCreg</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ABCreg(output_folder, S, P, tol, abcreg)</code></pre><p>Performing ABC inference using ABCreg. Please, be sure the input folder contain the observed and summary statistics produc  ed by MKtest.summary_statistics().</p><p><strong>Arguments</strong></p><ul><li><code>output_folder::String</code> : folder containing the observed data and summary estatistics. It will be used to output the posterior distributions</li><li><code>P::Int64</code> : number of parameters to infer.</li><li><code>S::Int64</code> : number of summary stastitics to perform the inference.</li><li><code>tol::Float64</code> : tolerance value. It defines the number of accepted values at ABC inference.</li><li><code>abcreg::String</code> : path to ABCreg binary.</li></ul><p><strong>Output</strong></p><ul><li><code>posteriors:Vector{Matrix{Float64}}</code>: posteriors distributions estimated by ABCreg.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/abc.jl#L17-L30">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.summary_abc" href="#MKtest.summary_abc"><code>MKtest.summary_abc</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">summary_abc(posteriors,stat)</code></pre><p>Posterior distributions statistics. The function estimates Min., 0.5% Perc., Median, Mean, Mode, &quot;99.5% Perc., Max values. The mode is estimated following R package abc mode function. The argument <code>stat</code> define the output estimation.</p><p><strong>Arguments</strong></p><ul><li><code>posterior::Matrix{Float64}</code> : posterior distribution.</li><li><code>stat::String</code>: output estimation.</li><li><code>plot_path{String,Nothing}</code>: path to save posterior plot.</li></ul><p><strong>Output</strong></p><ul><li><code>DataFrame</code>: posterior statistics.</li><li><code>DataFrame</code>: chosen statistic inference.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/abc.jl#L141-L152">source</a></section></article><h2 id="MKtest"><a class="docs-heading-anchor" href="#MKtest">MKtest</a><a id="MKtest-1"></a><a class="docs-heading-anchor-permalink" href="#MKtest" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.aMK" href="#MKtest.aMK"><code>MKtest.aMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">aMK(param,α;dac_rm,na_rm)</code></pre><p>Function to estimate the asymptotic value of α(x).</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model</li><li><code>α::Matrix</code>: α(x) values.</li><li><code>dac_rm::Bool</code>: Remove any value not selected at param.dac.</li><li><code>na_rm::Bool</code>: Remove any NA value at α(x)</li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{Float64, Vector{Float64}, Vector{Float64}}</code>: Tuple containing aMK estimation, CI estimation, model output.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/methods.jl#L50-L63">source</a></section><section><div><pre><code class="nohighlight hljs">aMK(param,α;dac_rm,na_rm)</code></pre><p>Function to estimate the asymptotic value of α(x).</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model</li><li><code>α::Vector</code>: Multiple α(x) vectors.</li><li><code>dac_rm::Bool</code>: Remove any value not selected at param.dac.</li><li><code>na_rm::Bool</code>: Remove any NA value at α(x)</li></ul><p><strong>Returns</strong></p><ul><li><code>Vector{Tuple}</code>: Tuple containing aMK estimation, CI estimation, model output.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/methods.jl#L113-L126">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.imputedMK</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="MKtest.fwwMK" href="#MKtest.fwwMK"><code>MKtest.fwwMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fwwMK(param,sfs,divergence;m,cutoff)</code></pre><p>Function to estimate the fwwMK (https://doi.org/10.1038/4151024a).</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model.</li><li><code>sfs::Vector</code>: SFS data parsed from parse_sfs().</li><li><code>divergence::Vector</code>: divergence data from parse_sfs().</li><li><code>cutoff{Float64}</code>: frequency cutoff to perform fwwMK.</li><li><code>m::Matrix{Int64}</code>: total number of sites parsed from parse_sfs()</li></ul><p><strong>Returns</strong></p><ul><li>`Dict: Dictionary containing imputedMK estimation.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/methods.jl#L361-L374">source</a></section><section><div><pre><code class="nohighlight hljs">fwwMK(param,sfs,divergence;m,cutoff)</code></pre><p>Function to estimate the fwwMK (https://doi.org/10.1038/4151024a).</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the variables required to solve the model.</li><li><code>sfs::Matrix</code>: SFS data parsed from parse_sfs().</li><li><code>divergence::Matrix</code>: divergence data from parse_sfs().</li><li><code>cutoff{Float64}</code>: frequency cutoff to perform fwwMK.</li><li><code>m::Matrix{Int64}</code>: total number of sites parsed from parse_sfs()</li></ul><p><strong>Returns</strong></p><ul><li>`Dict: Dictionary containing imputedMK estimation.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/methods.jl#L428-L441">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.standardMK" href="#MKtest.standardMK"><code>MKtest.standardMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">standardMK(sfs,divergence;m)</code></pre><p>Function to estimate the original α value.</p><p><strong>Arguments</strong></p><ul><li><code>sfs::Matrix{Float64}</code>: SFS array</li><li><code>divergence::Array</code>: divegence count array</li><li><code>m::Union{Nothing,Array}</code>: non-synonymous and synonymous sites# Returns</li></ul><p><strong>Output</strong></p><ul><li>`Dict: Dictionary containing results</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/91bce9c29c28117d148d5f4a1cde390fca622194/src/methods.jl#L492-L504">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.grapes</code>. Check Documenter&#39;s build log for details.</p></div></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../bootstrap/">« Bootstrap</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.13 on <span class="colophon-date" title="Wednesday 20 September 2023 14:04">Wednesday 20 September 2023</span>. Using Julia version 1.9.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
