<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reference · MKtest.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://jmurga.github.io/MKtest.jl/reference/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">MKtest.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Package Overview</span><ul><li><a class="tocitem" href="../analytical/">Analytical estimations</a></li><li><a class="tocitem" href="../data/">Processing data</a></li><li><a class="tocitem" href="../mk/">MK approaches</a></li></ul></li><li><span class="tocitem">Infering the rate and strength of adaptation</span><ul><li><a class="tocitem" href="../empirical/">Empirical estimation</a></li><li><a class="tocitem" href="../rates/">Rates</a></li><li><a class="tocitem" href="../input/">Input data</a></li><li><a class="tocitem" href="../summstat/">Summary statistics</a></li><li><a class="tocitem" href="../abc/">ABC inference</a></li><li><a class="tocitem" href="../multiple/">Multiple dataset</a></li></ul></li><li><a class="tocitem" href="../cli/">Command-Line interface</a></li><li class="is-active"><a class="tocitem" href>Reference</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#MKtest-estimation"><span>MKtest estimation</span></a></li><li><a class="tocitem" href="#Fixations"><span>Fixations</span></a></li><li><a class="tocitem" href="#Polymorphism"><span>Polymorphism</span></a></li><li><a class="tocitem" href="#Rates"><span>Rates</span></a></li><li><a class="tocitem" href="#Summary-statistics"><span>Summary statistics</span></a></li><li><a class="tocitem" href="#Inference-tools"><span>Inference tools</span></a></li><li><a class="tocitem" href="#MKtest"><span>MKtest</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Reference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jmurga/MKtest.jl/blob/master/docs/src/reference.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Model-parameters"><a class="docs-heading-anchor" href="#Model-parameters">Model parameters</a><a id="Model-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Model-parameters" title="Permalink"></a></h1><p><em>adap</em> is the only variable exported from <em>MKtest</em> module. It is a Mutable structure contaning the variables required to solve the analytical approach. Any value can be easly changed. Remember <em>adap</em> should be change before the execution, in other case, <span>$\alpha_{(x)}$</span> will be solve with the default values. To change all the values at once, you can use <a href="#MKtest.parameters"><code>MKtest.parameters</code></a> in order to set specific models in a new variable.</p><article class="docstring"><header><a class="docstring-binding" id="MKtest.parameters" href="#MKtest.parameters"><code>MKtest.parameters</code></a> — <span class="docstring-category">Type</span></header><section><div><p>Mutable structure containing the variables required to solve the analytical approach. All the functions are solve using the internal values of the structure. You should declare a mutable structure to the perform the analytical estimations.</p><p><strong>Parameters</strong></p><ul><li><code>gam_neg::Int64</code>: Selection coefficient for deleterious alleles</li><li><code>gL::Int64</code>: Selection coefficient for weakly benefical alleles</li><li><code>gH::Int64</code>: Selection coefficient for strongly benefical alleles</li><li><code>al_low::Float64</code>: Proportion of α due to weak selection</li><li><code>al_tot::Float64</code>: α</li><li><code>θ_noncoding::Float64</code>: Mutation rate defining BGS strength</li><li><code>θ_coding::Float64</code>: Mutation rate on coding region</li><li><code>al::Float64</code>: DFE shape parameter </li><li><code>be::Float64</code>: DFE scale parameter</li><li><code>B::Float64</code>: BGS strength</li><li><code>B_bins::Array{Float64,1}</code>: BGS values to simulate</li><li><code>ppos_l::Float64</code>: Fixation probabily of weakly beneficial alleles</li><li><code>ppos_h::Float64</code>: Fixation probabily of strongly beneficial alleles</li><li><code>N::Int64</code>: Population size</li><li><code>n::Int64</code>: Sample size</li><li><code>Lf::Int64</code>: Flanking region length</li><li><code>ρ::Float64</code>: Recombination rate</li><li><code>TE::Float64</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L6-L29">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.Br" href="#MKtest.Br"><code>MKtest.Br</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Br(Lmax,theta)</code></pre><p>Expected reduction in nucleotide diversity. Explored at <a href="https://doi.org/10.1017/S0016672300032365">Charlesworth B., 1994</a>:</p><p class="math-container">\[\frac{\pi}{\pi_{0}} = e^{\frac{4\muL}{2rL+t}}\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li><li><code>theta::Float64</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected reduction in diversity given a non-coding length, mutation rate and defined recombination.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L71-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.set_θ!" href="#MKtest.set_θ!"><code>MKtest.set_θ!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">set_θ!(param)</code></pre><p>Find the optimum mutation given the expected reduction in nucleotide diversity (B value) in a locus.</p><p><strong>Returns</strong></p><ul><li><code>adap.θ_noncoding::Float64</code>: changes adap.θ_noncoding value.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L98-L105">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.set_ppos!" href="#MKtest.set_ppos!"><code>MKtest.set_ppos!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">set_ppos!(param)</code></pre><p>Find the probabilty of positive selected alleles given the model. It solves a equation system taking into account fixations probabilities of weakly and strong beneficial alleles.</p><p><strong>Returns</strong></p><ul><li><code>Tuple{Float64,Float64}</code>: weakly and strong beneficial alleles probabilites.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L127-L134">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.binom_op!" href="#MKtest.binom_op!"><code>MKtest.binom_op!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">binom_op(param)</code></pre><p>Binomial convolution to sample the allele frequencies probabilites depending on background selection values and sample size.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li><li><code>binom::binomialDict</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Float64,2}</code>: convoluted SFS for each B value defined in the model (param.B_bins). The estimations are saved at <em>convolutedBn.bn</em>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L155-L166">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.Φ" href="#MKtest.Φ"><code>MKtest.Φ</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Φ(param,gamma,ppos)</code></pre><p>Reduction in fixation probabilty due to background selection and linkage. The formulas used have been subjected to several theoretical works (<a href="https://doi.org/10.1017/S0016672300032365">Charlesworth B., 1994</a>, <a href="https://www.genetics.org/content/141/4/1605">Hudson et al., 1995</a>, <a href="https://doi.org/10.1017/S0016672300033619">Nordborg et al. 1995</a>, <a href="https://www.genetics.org/content/140/2/821">Barton NH., 1995</a>).</p><p>The fixation probabilty of selected alleles are reduce by a factor <span>$\phi$</span>:</p><p class="math-container">\[\phi(t,s) = e^{[\frac{-2\mu}{t(1+\frac{rL}{t}+\frac{2s}{t})}]}\]</p><p>Multiplying across all deleterious linkes sites, we find:</p><p class="math-container">\[\Phi = \prod_{1}^{L} = \phi(t,s) = e^{\frac{-2t\mu(\psi[1,\frac{r+2s+L}{r}] - \psi[1,\frac{r(L+1)+2s+t}{r}])}{r^2}}\]</p><p class="math-container">\[\phi(t,s) = e^{\frac{-2t\mu(\psi[1,\frac{r+2s+L}{r}] - \psi[1,\frac{r(L+1)+2s+t}{r}])}{r^2}}\]</p><p><strong>Arguments</strong></p><ul><li><code>gamma::Int64</code>: selection coefficient.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixations under background selection.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L208-L237">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.analytical_alpha" href="#MKtest.analytical_alpha"><code>MKtest.analytical_alpha</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">analyticalAlpha(param, binom)</code></pre><p>Analytical α(x) estimation. Solve α(x) generally. We used the expected rates of divergence and polymorphism to approach the asympotic value accouting for background selection, weakly and strong positive selection. α(x) can be estimated taking into account the role of positive selected alleles or not. In this way we explore the role of linkage to deleterious alleles in the coding region.</p><p class="math-container">\[\mathbb{E}[\alpha_{x}] =  1 - \left(\frac{\mathbb{E}[D_{s}]}{\mathbb{E}[D_{N}]}\frac{\mathbb{E}[P_{N}]}{\mathbb{E}[P_{S}]}\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Float64,1}</code> α(x).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/parameters.jl#L252-L265">source</a></section></article><h1 id="MKtest-estimation"><a class="docs-heading-anchor" href="#MKtest-estimation">MKtest estimation</a><a id="MKtest-estimation-1"></a><a class="docs-heading-anchor-permalink" href="#MKtest-estimation" title="Permalink"></a></h1><h2 id="Fixations"><a class="docs-heading-anchor" href="#Fixations">Fixations</a><a id="Fixations-1"></a><a class="docs-heading-anchor-permalink" href="#Fixations" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.fix_neut" href="#MKtest.fix_neut"><code>MKtest.fix_neut</code></a> — <span class="docstring-category">Function</span></header><section><div><ul><li>fix_neut()</li></ul><p>Expected neutral fixations rate reduce by B value.</p><p class="math-container">\[\mathbb{E}[D_{s}] = (1 - p_{-} - p_{+}) B \frac{1}{2N}\]</p><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of neutral fixations.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/fixations.jl#L8-L20">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.fix_neg_b</code>. Check Documenter&#39;s build log for details.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>MKtest.p_fix</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="MKtest.fix_pos_sim" href="#MKtest.fix_pos_sim"><code>MKtest.fix_pos_sim</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fix_pos_sim(gamma,ppos)</code></pre><p>Expected positive fixations rate reduced due to the impact of background selection and linkage. The probabilty of fixation of positively selected alleles is reduced by a factor Φ across all deleterious linked sites <a href="@ref"><code>Analytical.Φ</code></a>.</p><p class="math-container">\[\mathbb{E}[D_{n+}] =  \Phi \cdot \mathbb{E}[D_{n+}]\]</p><p><strong>Arguments</strong></p><ul><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixations under background selection</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/fixations.jl#L89-L105">source</a></section></article><h2 id="Polymorphism"><a class="docs-heading-anchor" href="#Polymorphism">Polymorphism</a><a id="Polymorphism-1"></a><a class="docs-heading-anchor-permalink" href="#Polymorphism" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.sfs_neut" href="#MKtest.sfs_neut"><code>MKtest.sfs_neut</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sfs_neut()</code></pre><p>Expected rate of neutral allele frequency reduce by backgrou	nd selection. The spectrum depends on the number of individual []</p><p class="math-container">\[\mathbb{E}[Ps_{(x)}] = \sum{x^{*}=x}{x^{*}=1}f_{B}(x)\]</p><p><strong>Input</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model</li><li><code>SparseMatrixCSC{Float64, Int64}</code>: convoluted SFS given the B value.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected rate of neutral alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/polymorphism.jl#L10-L24">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.sfs_pos" href="#MKtest.sfs_pos"><code>MKtest.sfs_pos</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sfs_pos(s,p)</code></pre><p>Expected rate of positive selected allele frequency reduce by background selection. The spectrum depends on the number of individuals.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model.</li><li><code>s::Int64</code>: selection strength.</li><li><code>p::Float64</code>: positive selected alleles probabilty.</li><li><code>binom::SparseMatrixCSC{Float64, Int64}</code>: convoluted SFS given the B value.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected positive selected alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/polymorphism.jl#L43-L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.sfs_neg" href="#MKtest.sfs_neg"><code>MKtest.sfs_neg</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sfs_neg(param,p)</code></pre><p>Expected rate of positive selected allele frequency reduce by background selection. Spectrum drawn on a gamma DFE. It depends on the number of individuals.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model</li><li><code>p::Float64</code>: positive selected alleles probabilty.</li><li><code>binom::SparseMatrixCSC{Float64, Int64}</code>: convoluted SFS given the B value.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected negative selected alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/polymorphism.jl#L123-L135">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.cumulative_sfs" href="#MKtest.cumulative_sfs"><code>MKtest.cumulative_sfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cumulative_sfs(sfs_tmp)</code></pre><p>Changing SFS considering all values above a frequency <em>x</em>. The original asymptotic-MK approach takes Pn(x) and Ps(x) as the number of polymorphic sites at frequency <em>x</em> rather than above <em>x</em>, but this approach scales poorly as sample size increases. We define the polymorphic spectrum as stated above since these quantities trivially have the same asymptote but are less affected by changing sample size.</p><p><strong>Arguments</strong></p><ul><li><code>sfs_tmp::Vector{Float64}</code>: SFS array</li><li><code>freqs::Bool</code>: true/false wheter the input SFS array containing a first columns with the frequencies.</li></ul><p><strong>Output</strong></p><ul><li><code>sfs_tmp::Vector{Float64}</code>: SFS array</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/polymorphism.jl#L162-L172">source</a></section></article><h2 id="Rates"><a class="docs-heading-anchor" href="#Rates">Rates</a><a id="Rates-1"></a><a class="docs-heading-anchor-permalink" href="#Rates" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.rates" href="#MKtest.rates"><code>MKtest.rates</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">rates(param::parameters,iterations::Int64,divergence::Array,sfs::Array)</code></pre><p>Function to solve randomly <em>N</em> scenarios. The function will create <em>N</em> models, defined by <code>Analytical.parameters()</code>, to estimate analytically fixation and polymorphic rates for each model. The rates will be used to compute summary statistics required at ABC. The function output a HDF5 file containing the solved models, the selected DAC and the analytical rates. </p><p>If ρ and/or theta are set to <code>nothing</code>, the function will input random values given the range 0.0005:0.0005:0.01. Otherwise you can fix the values.</p><p>If gL is set to <code>nothing</code>, the function will not account the role of the weakly selected alleles in the estimation.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: mutable structure containing the model</li><li><code>binom::binomialDict</code> : structure containing the binomial convolution</li><li><code>gH::Array{Int64,1}</code> : Range of strong selection coefficients</li><li><code>gL::Union{Array{Int64,1},Nothing}</code>: Range of weak selection coefficients</li><li><code>gam_neg::Array{Int64,1}</code> : Range of deleterious selection coefficients</li><li><code>theta::Union{Float64,Nothing}</code> : Population-scaled mutation rate on coding region</li><li><code>ρ::Union{Float64,Nothing}</code> : Population-scaled recombination rate</li><li><code>shape::Float64=0.184</code> : DFE shape parameter</li><li><code>iterations::Int64</code> : Number of solutions</li><li><code>output::String</code> : File to output HDF5 file</li></ul><p><strong>Returns</strong></p><ul><li><code>Array</code>: summary statistics</li><li><code>Output</code>: HDF5 file containing models solved and rates.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/rates.jl#L1-L24">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.iter_rates" href="#MKtest.iter_rates"><code>MKtest.iter_rates</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">iter_rates(param::parameters,afac::Float64,bfac::Float64,al_tot::Float64,al_low::Float64,divergence::Array,sfs::Array)</code></pre><p>Estimating rates given a model for all B range.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li><li><code>binom::Dict</code></li><li><code>al_tot::Float64</code></li><li><code>al_low::Float64</code></li><li><code>gH::Int64</code></li><li><code>gL::Int64</code></li><li><code>gam_neg::Int64</code></li><li><code>afac::Float64</code></li><li><code>ρ::Float64</code></li><li><code>θ::Float64</code></li></ul><p><strong>Output</strong></p><ul><li><code>Array{Float64,2}</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/rates.jl#L123-L141">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.getting_rates" href="#MKtest.getting_rates"><code>MKtest.getting_rates</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">getting_rates(gammaL,gammaH,ppos_l,ppos_h,observedData,nopos)</code></pre><p>Estimating analytical rates of fixation and polymorphism to approach α value accouting for background selection, weakly and strong positive selection. Output values will be used to sample from a Poisson distribution the total counts of polymorphism and divergence using observed data. </p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code></li><li><code>binom::SparseMatrixCSC{Float64,Int64}</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Float64,2}</code> containing solved model, fixation and polymorphic rates</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/rates.jl#L178-L188">source</a></section></article><h2 id="Summary-statistics"><a class="docs-heading-anchor" href="#Summary-statistics">Summary statistics</a><a id="Summary-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Summary-statistics" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.poisson_fixation" href="#MKtest.poisson_fixation"><code>MKtest.poisson_fixation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">poisson_fixation(observed_values,λds, λdn)</code></pre><p>Divergence sampling from Poisson distribution. The expected neutral and selected fixations are subset through their relative expected rates (<a href="#MKtest.fix_neut"><code>fix_neut</code></a>, <a href="@ref"><code>fix_neg_b</code></a>, <a href="#MKtest.fix_pos_sim"><code>fix_pos_sim</code></a>). Empirical values are used are used to simulate the locus <em>L</em> along a branch of time <em>T</em> from which the expected <em>Ds</em> and <em>Dn</em> raw count estimated given the mutation rate (<span>$\mu$</span>). Random number generation is used to subset samples arbitrarily given the success rate <span>$\lambda$</span> in the distribution.</p><p class="math-container">\[\mathbb{E}[D_N] = X \in Poisson\left(\lambda = D \times \left[\frac{\mathbb{E}[D_+] + \mathbb{E}[D_-]}{\mathbb{E}[D_+] + \mathbb{E}[D_-] + \mathbb{E}[D_0]}\right]\right)\]</p><p class="math-container">\[\mathbb{E}[D_S] = X \in Poisson\left(\lambda = D \times \left[\frac{\mathbb{E}[D_0]}{\mathbb{E}[D_+] + \mathbb{E}[D_-] + \mathbb{E}[D_0]}\right]\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>observed_values::Array</code>: Array containing the total observed divergence.</li><li><code>λds::Float64</code>: expected neutral fixations rate.</li><li><code>λdn::Float64</code>: expected selected fixations rate.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Int64,1}</code> containing the expected count of neutral and selected fixations.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/summary_statistics.jl#L4-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.poisson_polymorphism" href="#MKtest.poisson_polymorphism"><code>MKtest.poisson_polymorphism</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">poisson_polymorphism(observed_values,λps,λpn)</code></pre><p>Polymorphism sampling from Poisson distributions. The total expected neutral and selected polimorphism are subset through the relative expected rates at the frequency spectrum (<a href="#MKtest.fix_neut"><code>fix_neut</code></a>, <a href="#MKtest.sfs_neut"><code>sfs_neut</code></a>,). Empirical sfs are used to simulate the locus <em>L</em> along a branch of time <em>T</em> from which the expected <em>Ps</em> and <em>Pn</em> raw count are estimated given the mutation rate (<span>$\mu$</span>). Random number generation is used to subset samples arbitrarily from the whole sfs given each frequency success rate <span>$\lambda$</span> in the distribution.</p><p>The success rate managing the Poisson distribution by the observed count each frequency.  We considered both sampling variance and process variance is affecting the number of variable alleles we sample from SFS. This variance arises from the random mutation-fixation process along the branch. To incorporate this variance we do one sample per frequency-bin and use the total sampled variation and the SFS for the summary statistics.</p><p class="math-container">\[\mathbb{E}[P_N] = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)} \times \left[\frac{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}]}{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}] + \mathbb{E}[P_{0(x)}]}\right]\right)\]</p><p class="math-container">\[\mathbb{E}[P_S] = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)} \times \left[\frac{\mathbb{E}[P_{0(x)}]}{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}] + \mathbb{E}[P_{0(x)}]}\right]\right)\]</p><p><strong>Arguments</strong></p><ul><li><code>observed_values::Array{Int64,1}</code>: Array containing the total observed divergence.</li><li><code>λps::Array{Float64,1}</code>: expected neutral site frequency spectrum rate.</li><li><code>λpn::Array{Float64,1}</code>: expected selected site frequency spectrum rate.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Int64,2}</code> containing the expected total count of neutral and selected polymorphism.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/summary_statistics.jl#L42-L64">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.sampled_from_rates" href="#MKtest.sampled_from_rates"><code>MKtest.sampled_from_rates</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sampled_alpha(observed_values,λds, λdn)</code></pre><p>Ouput the expected values from the Poisson sampling process. Please check <a href="#MKtest.poisson_fixation"><code>poisson_fixation</code></a> and <a href="#MKtest.poisson_polymorphism"><code>poisson_polymorphism</code></a> to understand the samplingn process. α(x) is estimated through the expected values of Dn, Ds, Pn and Ps.</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code>: Array containing the total observed divergence.</li><li><code>d::Array</code>: observed divergence.</li><li><code>afs::Array</code>: observed polymorphism.</li><li><code>λdiv::Array{Float64,2}</code>: expected fixations rate.</li><li><code>λdiv::Array{Float64,2}</code>: expected site frequency spectrum rates.</li></ul><p><strong>Returns</strong></p><p>α<em>summaries,exp</em>dn,exp<em>ds,exp</em>pn,exp_ps,ssAlpha</p><ul><li><code>Array{Int64,2}</code> containing α(x) values.</li><li><code>Array{Int64,1}</code> expected non-synonymous divergence.</li><li><code>Array{Int64,1}</code> expected synonymous divergence.</li><li><code>Array{Int64,1}</code> expected non-synonymous polymorphism.</li><li><code>Array{Int64,1}</code> expected synonymous polymorphism.</li><li><code>Array{Int64,1}</code> expected synonymous polymorphism.</li><li><code>Array{Int64,1}</code> expected synonymous polymorphism.</li><li><code>Array{Int64,2}</code> containing α(x) binned values.</li></ul><pre><code class="nohighlight hljs">sampled_from_rates(gammaL,gammaH,ppos_l,ppos_h,observedData,nopos)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/summary_statistics.jl#L84-L108">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.summary_statistics" href="#MKtest.summary_statistics"><code>MKtest.summary_statistics</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">summary_statistics(param::parameters,rates::JLD2.JLDFile,analysis_folder::String,summstat_size::Int64,replicas::Int64,bootstrap::Bool)</code></pre><p>Estimate summary statistics using observed data and analytical rates. <em>analysis_folder</em> will check for the SFS and divergence file and will be used to output summary statistics</p><p><strong>Arguments</strong></p><ul><li><code>param::parameters</code> : Mutable structure containing the models</li><li><code>h5_file::String</code> : HDF5 containing solved models, fixation and polymorphic rates</li><li><code>sfs::Vector{Matrix{Float64}}</code>: SFS matrix.</li><li><code>divergence::Vector{Matrix{Float64}}</code> : divergence matrix.</li><li><code>summstat_size::Int64</code> : Number of summary statistics</li><li><code>replicas::Int64</code> : Number of bootstrap replicas</li><li><code>bootstrap::Bool</code> : Boolean to perform or not bootstrapping</li></ul><p><strong>Output</strong></p><ul><li>Summary statistics to ABC inference</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/summary_statistics.jl#L134-L149">source</a></section></article><h2 id="Inference-tools"><a class="docs-heading-anchor" href="#Inference-tools">Inference tools</a><a id="Inference-tools-1"></a><a class="docs-heading-anchor-permalink" href="#Inference-tools" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.parse_sfs" href="#MKtest.parse_sfs"><code>MKtest.parse_sfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">parse_sfs(;sample_size::Int64,data::String)</code></pre><p>Function to parse polymorphism and divergence by subset of genes. The input data is based on supplementary material described at <a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. 2019</a>. Please be sure the file is tabulated.</p><table><tr><th style="text-align: right">GeneId</th><th style="text-align: right">Pn</th><th style="text-align: right">DAF seppareted by commas</th><th style="text-align: right">Ps</th><th style="text-align: right">DAF separated by commas</th><th style="text-align: right">Dn</th><th style="text-align: right">Ds</th></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr></table><p><strong>Arguments</strong></p><ul><li><code>sample_size::Int64</code>: data sample size.</li><li><code>data</code>: File containing polymorphic and divergence data</li><li><code>gene_list</code>: File containing gene IDs to subset. You can perform multiple subset input a file where eachrow contain the gene IDs to subset.</li><li><code>sfs_columns::Array{Int64,1}</code>: non-synonymous and synonymous daf columns. Please introduce first the non-synonymous number.</li><li><code>div_columns::Array{Int64,1}</code>: non-synonymous and synonymous divergence columns. Please introduce first the non-synonymous number.</li><li><code>bins::Array{Int64,1}</code>: bin to collapse the SFS from <code>sample_size</code> to new <code>bin</code> size</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Float64,1}</code>: α values</li><li><code>Array{Float64,2}</code>: Site Frequency Spectrum</li><li><code>Array{Float64,1}</code>: Synonymous and non-synonymous divergence counts</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/infer_tools.jl#L1-L24">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.ABCreg" href="#MKtest.ABCreg"><code>MKtest.ABCreg</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ABCreg(analysis_folder, replicas, S, tol, abcreg)</code></pre><p>Performing ABC inference using ABCreg. Please, be sure your analysis_folder contain the files alphas.txt and summaries.txt produced by Analytical.summaryStatsFromRates()</p><p><strong>Arguments</strong></p><ul><li><code>analysis_folder::String</code> : Folder containing the observed data and summary estatistics. It will be used to output the posterior distributions</li><li><code>P::Int64</code> : Number of parameters to perform the inference.</li><li><code>S::Int64</code> : Number of summary stastitics to perform the inference.</li><li><code>tol::Float64</code> : Tolerance value. It define the number of accepted values at ABC inference</li><li><code>abcreg::String</code> : Path to ABCreg binary</li></ul><p><strong>Output</strong></p><p>Files containing posterior distributions from ABCreg</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/infer_tools.jl#L93-L108">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.source_plot_map_r" href="#MKtest.source_plot_map_r"><code>MKtest.source_plot_map_r</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Function to download and source plotMap function. We do not include at part of the module to avoid external dependecies. Once the function is execute properly you will have a function called *plotMap which used R to 
	estimate and plot the Maximum A Posterior following ABCreg example. It uses locfit and ggplot2 libraries.</code></pre><p>#Arguments</p><ul><li><code>script_path::String</code>: Path and name to save MAP scritp.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/infer_tools.jl#L170-L175">source</a></section></article><h2 id="MKtest"><a class="docs-heading-anchor" href="#MKtest">MKtest</a><a id="MKtest-1"></a><a class="docs-heading-anchor-permalink" href="#MKtest" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="MKtest.aMK" href="#MKtest.aMK"><code>MKtest.aMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">aMK(x,columns)</code></pre><p>Function to estimate the asymptotic value of α(x).</p><p><strong>Arguments</strong></p><ul><li><code>alpha_values::Array{Float64,1}</code>: α(x) array.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Float64,2}</code>: Array of array containing asymptotic values, lower confidence interval, higher confidence interval</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/methods.jl#L1-L10">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.imputedMK" href="#MKtest.imputedMK"><code>MKtest.imputedMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">imputedMK(sfs,divergence,m,cutoff)</code></pre><p>Function to estimate the imputedMK</p><p><strong>Arguments</strong></p><ul><li><code>sfs::Matrix{Float64}</code>: SFS array</li><li><code>divergence::Array</code>: divegence count array</li><li><code>m::Union{Nothing,Array}</code>: non-synonymous and synonymous sites</li><li><code>cutoff{Float64}</code>: frequency cutoff to perform imputedMK</li></ul><p><strong>Returns</strong></p><ul><li>`Dict: Dictionary containing results</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/methods.jl#L30-L42">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.fwwMK" href="#MKtest.fwwMK"><code>MKtest.fwwMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fwwMK(sfs,divergence,m,cutoff)</code></pre><p>Function to estimate the Fay, Wycoff and Wu&#39;s MK extension.</p><p><strong>Arguments</strong></p><ul><li><code>sfs::Matrix{Float64}</code>: SFS array</li><li><code>divergence::Array</code>: divegence count array</li><li><code>m::Union{Nothing,Array}</code>: non-synonymous and synonymous sites</li><li><code>cutoff{Float64}</code>: frequency cutoff to perform fwwMK</li></ul><p><strong>Returns</strong></p><ul><li>`Dict: Dictionary containing results</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/methods.jl#L102-L114">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MKtest.standardMK" href="#MKtest.standardMK"><code>MKtest.standardMK</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">standardMK(x,columns)</code></pre><p>Function to estimate the original α value</p><p><strong>Arguments</strong></p><ul><li><code>sfs::Matrix{Float64}</code>: SFS array</li><li><code>divergence::Array</code>: divegence count array</li><li><code>m::Union{Nothing,Array}</code>: non-synonymous and synonymous sites# Returns</li></ul><p><strong>Output</strong></p><ul><li>`Dict: Dictionary containing results</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/MKtest.jl/blob/f184d391b2b785c1ae5875b0832661b685616ed9/src/methods.jl#L154-L166">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../cli/">« Command-Line interface</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.13 on <span class="colophon-date" title="Wednesday 4 May 2022 11:25">Wednesday 4 May 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
